File: 프로그래머스/2/42577. 전화번호 목록/전화번호 목록.cpp
#include <string>
#include <vector>
#include <iostream>
#include <set>

using namespace std;

bool solution(vector<string> phone_book) {
    
    set<string> s; // 전화번호 저장
    for(int i=0 ; i<phone_book.size(); i++){
        s.insert(phone_book[i]);
    }
    
    for(int i=0 ; i<phone_book.size(); i++){
        // 1. 각 전화번호 별로 가능한 모든 접두어 만들어서
        for(int j = 1; j < phone_book[i].size(); j++){
            string pf = phone_book[i].substr(0,j);
            // 2. 전화번호부에 있는지 확인
            if(s.find(pf) != s.end()){
                return false;
            }
        }
    }
    
    
    return true;
}

File: 프로그래머스/2/42577. 전화번호 목록/전화번호 목록.java
import java.util.*;
import java.io.*;

class Solution {
    public boolean solution(String[] phone_book) {
        
        Set<String> s = new HashSet<>();
        for(String number : phone_book){
            s.add(number);
        }
        
        for(String number: phone_book){
            for(int i=1; i<number.length(); i++){
                String pf = number.substring(0,i);
                if(s.contains(pf)){
                    return false;
                }
            }
        }
        
        return true;
    }
}